### ISSUE TITLE
Send heartbeats when not receiving heartbeats frames or publishing

### ISSUE THREAD

**Issue by llschloesser** [CopernicaMarketingSoftware/AMQP-CPP#137]
As I understand RabbitMQ, any data counts as a heartbeat and for that reason, if a client is steadily consuming data from the server, the server will never send an explicit heartbeat frame because it views any data sent as a valid heartbeat. amqp-cpp is written such that explicit heartbeats are sent only when explicit heartbeats are received. Because of this design, a pure consumer client will have its connection terminated by the server (connection reset by peer) with RabbitMQ logs showing the reason being a heartbeat timeout due to no data received from the client.

Is it possible to embed a timer into the amp-cpp library to recognize when no data has been sent for the negotiated timeout period and then stimulate the transmission of a heartbeat frame? I can accomplish this in my application code, but I would expect that this is a common situation and might be best handled by the amp-cpp library.

**Comment by EmielBruijntjes** [comment]
AFAIK, the heartbeats are sent by the server to check if the client is alive, and the client just has to reply to them. If you keep your event loop active, the AMQP-CPP library immediately sends back the heartbeat-reply, so the connection always stay open.

Do you have an example in which a pure consumer loses its connection?

**Comment by llschloesser** [comment]
I am currently using boost::asio and my own connection handler that won't be too minimal. I will create a minimal example using AMQP::TcpConnection and post it here. The broker version I'm using is 3.6.3.

See [this page](https://www.rabbitmq.com/heartbeats.html) for the suggestion that heartbeats won't be sent by the server so long as messages are being consumed:

> Any traffic (e.g. protocol operations, published messages, acknowledgements) counts for a valid heartbeat. Clients may choose to send heartbeat frames regardless of whether there was any other traffic on the connection but some only do it when necessary.

It's important to add that in my use case we are not acking. If we were to ack, then I believe (I will test this) we would not observe this issue as the ack would count as a heartbeat.

**Comment by EmielBruijntjes** [comment]
My interpretation was that the server initiates them when it finds that the client has been idle for too long, but I can see your point too. We need to test what is going on.

I will check this with an open connection that is not used at all.

**Comment by fantastory** [comment]
> This value is negotiated between the client and RabbitMQ server at the time of connection. The client must be configured to request heartbeats.

They say the value is negotiated, so indeed server could drop connection if there is no data from client by two intervals.

**Comment by EmielBruijntjes** [comment]
````
#include <iostream>
#include <amqpcpp.h>
#include <amqpcpp/libev.h>

class MyHandler : public AMQP::LibEvHandler
{
private:
    int _counter = 0;
    
public:
    MyHandler() : AMQP::LibEvHandler(EV_DEFAULT) {}
    
    virtual void onHeartbeat(AMQP::TcpConnection *connection) override
    {
        std::cout << "heartbeat " << ++_counter << " received" << std::endl;
    }
    
    virtual void onError(AMQP::TcpConnection *connection, const char *message) override
    {
        std::cout << "error " << message << std::endl;
    }

    virtual uint16_t onNegotiate(AMQP::TcpConnection *connection, uint16_t interval) override
    {
        std::cout << "interval " << interval << std::endl;
        
        // use very small heartbeat instead of the default one
        return 1;
    }
};

int main()
{
    MyHandler handler;
    AMQP::TcpConnection connection(&handler, AMQP::Address("amqp://guest:guest@localhost/"));
    
    ev_run(EV_DEFAULT);

    return 0;
}
````

The above example receives heartbeats from the server every second, and it does not seem to drop the connection (but I have only ran it for two minutes now).

**Comment by fantastory** [comment]
But does it receive any other data?
llschloesser is saying about situation when client is subscribed to a queue which constantly receives other data. 
I am not sure if it is theoretical discussion or does the RabbitMQ really behaves like this.

**Comment by EmielBruijntjes** [comment]
You can extend my example to show where it goes wrong. On my computer it now runs for 20 minutes or so without losing the connection.

**Comment by llschloesser** [comment]
fantastory has it right. If you don't receive messages from the server, it will send heartbeats, and so an 'idle' connection is preserved by back-and-forth heartbeats. It is only when the client is consuming messages, not receiving heartbeat frames, not acking (I tested this and acking preserves the connection), and never publishing messages that it will timeout due to never transmitting data to the server.  So to test this, the client which will be disconnected needs to be consuming, which means another client must be publishing.

**Comment by EmielBruijntjes** [comment]
````c++
#include <iostream>
#include <amqpcpp.h>
#include <amqpcpp/libev.h>
#include <unistd.h>

class MyHandler : public AMQP::LibEvHandler
{
private:
    // name of the handler
    std::string _name;
    
    // number of heartbeats received
    int _counter = 0;
    
public:
    // constructor
    MyHandler(const char *name) : AMQP::LibEvHandler(EV_DEFAULT), _name(name) {}
    
    // method that is called when a heartbeat is received
    virtual void onHeartbeat(AMQP::TcpConnection *connection) override
    {
        // debug output
        std::cout << _name << ": heartbeat " << ++_counter << " received" << std::endl;
    }
    
    // method called in case of an error
    virtual void onError(AMQP::TcpConnection *connection, const char *message) override
    {
        // debug output
        std::cout << _name << ": error " << message << std::endl;
    }

    // method to negotiate the heartbeat interval
    virtual uint16_t onNegotiate(AMQP::TcpConnection *connection, uint16_t interval) override
    {
        // debug output
        std::cout << _name << ": interval " << interval << std::endl;
        
        // use very small heartbeat instead of the default one
        return 1;
    }
};

int main()
{
    // handlers for connection level events
    MyHandler handler1("consumer");
    MyHandler handler2("publisher");
    
    // create two connections
    AMQP::TcpConnection connection1(&handler1, AMQP::Address("amqp://guest:guest@localhost/"));
    AMQP::TcpConnection connection2(&handler2, AMQP::Address("amqp://guest:guest@localhost/"));
    
    // create two channels, one on each connection
    AMQP::TcpChannel channel1(&connection1);
    AMQP::TcpChannel channel2(&connection2);
    
    // create a temporary queue on the first channel
    channel1.declareQueue(AMQP::autodelete | AMQP::exclusive).onSuccess([&channel1,&channel2](const std::string &name, uint32_t messages, uint32_t consumers) {
    
        // start consuming from the temporary queue
        channel1.consume(name, AMQP::noack).onReceived([&channel2,name](const AMQP::Message &message, uint64_t deliverytag, bool redelivered) {
            
            // wait a while
            sleep(1);
            
            // publish it too
            channel2.publish("", name, "");
        });

        // publish the first message
        channel2.publish("", name, "");
    });
    
    // run event loop
    ev_run(EV_DEFAULT);

    // done
    return 0;
}
````

Above example opens two connections: one that is only used for consuming, and one that is only used for publishing. RabbitMQ apparently sends out heartbeats on both these connections, and there is, also aparently, no need for the client to even initiate a heartbeat.

**Comment by llschloesser** [comment]
Perhaps the consumer is receiving heartbeats from the server because of the 1 second sleep? In that case the server hasn't sent data for 1 second and so recognizes that it needs to send a heartbeat frame, which is why the client is receiving heartbeats. If you take the program above and set the negotiated heartbeat timeout to something larger than 1 second (say, 5 secs to make is obvious), then my expectation is that the server will stop sending heartbeats, the client will then stop replying and the connection will be reset. I apologize I haven't had the time to get a simple example working, but I will try to soon.

To summarize, connection 1, if it is consuming messages more often than the heartbeat timeout, should not receive any heartbeats.

**Comment by EmielBruijntjes** [comment]
It is hard to believe, but you seem to be right. Both the client and the server must periodically initiate the heartbeat.
